### PR TITLE
Get new access tokens when expired, auto logout used if refresh token is expired/deleted, encrypt localStorage data (tokens)

### LINKED ISSUE #190: Handle package updates on Docker
Currently, if the `package-lock.json` is updated, simply running docker-compose doesn't install the new dependencies inside the docker container.

### LINKED ISSUE #210: Refresh Access Token when expired

### LINKED ISSUE #219: Websocket keeps trying to reconnect when not authenticated

### PR DESCRIPTION
**Pull request by deepansh96**
Fixes #210 
Fixes #190 
Fixes #219 

## Summary
- refreshes access token if it is expired given the refresh token has not expired
- logs the user out if refresh token has expired
- encrypts token in localStorage 

- Added a library `secure-ls` to encrypt while storing and decrypt while retrieving store data that's being saved to localStorage by `vuex-persistedstate`. More details [here](https://github.com/robinvdvleuten/vuex-persistedstate#encrypted-local-storage)
- Changed `VUE_APP_BACKEND_AUTH_URL` in `.env`. 
- Added two endpoints to `Endpoints.js`

   ```js
   convertTokenEndpoint = "/convert-token/"
   refreshTokenEndpoint = "/token/"
   ```
- This is done such that we get a base URL of ```http://0.0.0.0:8001/auth``` and add different auth endpoints on top of that. The two auth endpoints that we will use will be
    ```js
    VUE_APP_BACKEND_AUTH_URL + convertTokenEndpoint // for gauth token conversion
    VUE_APP_BACKEND_AUTH_URL + refreshTokenEndpoint // for getting a new access token
    ```
- For the refresh token logic, instead of implementing a new view function at the backend, I have just used the functionality offered by the **Social OAuth2** package that we're using for google auth. This saves us a custom implementation where we would've have to verify the refresh token, generate a new access token, revoke the old refresh token and other steps. How to use a refresh token to retrieve a new access token using **Social Oauth2** can be seen on the readme [here](https://github.com/RealmTeam/django-rest-framework-social-oauth2#testing-the-setup).
- To test in Postman - 
    ```js
    POST "http://0.0.0.0:8001/auth/token/"
    // request
    {
        "grant_type": "refresh_token",
        "client_id": "...",
        "client_secret": "...",
        "refresh_token": "2pmJHJCmHnGLSRlexJdQag1qA2i6Sj"
    }

    // response
    {
        "access_token": "WVLUSQkF4i51yNIWh4RKA1sK99QuSP",
        "expires_in": 86400,
        "token_type": "Bearer",
        "scope": "read write",
        "refresh_token": "HvOatAiMzL2hS5pnKbd2MO3XMc5NLA"
    }
    ```

- Added a new vuex store state variable - "loginStatus" (true or false). We'll watch for this variable in `App.vue` and if true, we'll log out the user and show them a dialog
- Added a dialog box to `App.vue` to show the popup "You've been logged out. Please log in again"
- The whole flow for handling 401 un authorized calls is inspired from [this](https://stackoverflow.com/questions/51563821/axios-interceptors-retry-original-request-and-access-original-promise) with some minor tweaks
- The case of handling 400 requests on the refreshTokenEndpoint itself is not mentioned in that post and I've implemented that on my own
- To test expiration of access_token, go to TablePlus and reduce the expiry date of the latest issued `access_token` and try to navigate on the platform. Everything should work fine
- To test expiration/revoked refresh_token, go to TablePlus, do the above step first, then go to the refresh_token table, make the entry in `access_token_id` column as `NULL` and enter some datetime in the column `revoked`. Basically we're emulating the revokation of a refresh token. Its important to emulate the expiry of the `access_token` along with this, otherwise the client will keep using the access_token for all requests. Expected behaviour - you will be logged out! 

## Test Plan
Tested locally.

### CODE REVIEW

File: src/store/index.js [hunk 1]
```diff
@@ -3,9 +3,22 @@ import createPersistedState from "vuex-persistedstate";
 import plioItems from "./modules/plioItems";
 import auth from "./modules/auth";
 import sync from "./modules/sync";
+import SecureLS from "secure-ls";
+
+// encrypt and decrypt the localStorage
+// github.com/robinvdvleuten/vuex-persistedstate#encrypted-local-storage
+var ls = new SecureLS({ isCompression: false });
```
Comment (dalmia) [inline src/store/index.js:10]:
```suggestion
var localStorage = new SecureLS({ isCompression: false });
```

File: src/store/index.js [hunk 2]
```diff
@@ -3,9 +3,22 @@ import createPersistedState from "vuex-persistedstate";
...
 
 export default createStore({
-  plugins: [createPersistedState()],
+  plugins: [
+    createPersistedState({
+      storage: {
+        getItem: (key) => ls.get(key),
+        setItem: (key, value) => ls.set(key, value),
+        removeItem: (key) => ls.remove(key),
```
Comment (dalmia) [inline src/store/index.js:18]:
```suggestion
        getItem: (key) => localStorage.get(key),
        setItem: (key, value) => localStorage.set(key, value),
        removeItem: (key) => localStorage.remove(key),
```

Comment (dalmia) [inline src/store/index.js:18]: Better to switch to storing to and fetching from cookies instead of local Storage?

File: src/pages/Player.vue [hunk 3]
```diff
@@ -433,7 +433,7 @@ export default {
     },
     handleQueryError(err) {
       // handles error encountered when fetching plio or creating new session
-      if (err.response && err.response.status == 404) {
+      if (err.response && [404, 400].includes(err.response.status)) {
```
Comment (dalmia) [inline src/pages/Player.vue:436]: should put the 404, 400 thing in a method

File: src/store/modules/auth.js [hunk 4]
```diff
@@ -7,6 +7,9 @@ const state = {
   user: null,
   activeWorkspace: "",
   userId: null,
+  isReAuthenticating: false,
+  reAuthenticationCall: null,
+  loginStatus: false,
```
Comment (dalmia) [inline src/store/modules/auth.js:12]: can use isAuthenticated - no need of this variable.

File: src/store/modules/auth.js [hunk 5]
```diff
@@ -7,6 +7,9 @@ const state = {
   user: null,
   activeWorkspace: "",
   userId: null,
+  isReAuthenticating: false,
+  reAuthenticationCall: null,
```
Comment (dalmia) [inline src/store/modules/auth.js:11]: reAuthenticationPromise?

File: src/services/API/User.js [hunk 6]
```diff
@@ -5,7 +5,10 @@ import {
   userFromTokenEndpoint,
   usersEndpoint,
   userConfigEndpoint,
+  convertTokenEndpoint,
+  refreshTokenEndpoint,
 } from "@/services/API/Endpoints.js";
+import store from "../../store";
```
Comment (dalmia) [inline src/services/API/User.js:11]:
```suggestion
import store from "@/store";
```

File: src/services/Functional/User.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
+import UserAPIService from "@/services/API/User.js";
+
+export default {
+  reAuthenticate(store) {
+    // return the stored reAuthCall if re-authentication is in process
```
Comment (dalmia) [inline src/services/Functional/User.js:5]:
```suggestion
    // return the stored promise from the re-authentication call if re-authentication is in process
```

File: src/services/Functional/User.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    if (store.state.auth.isReAuthenticating)
+      return store.state.auth.reAuthenticationCall;
+
+    // re authentication is in process
```
Comment (dalmia) [inline src/services/Functional/User.js:9]:
```suggestion
    // set re-authentication as being in process
```

File: src/services/Functional/User.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    store.dispatch("auth/setReAuthenticationState", true);
+
+    // start the refreshAccessToken call and save that promise in the variable
```
Comment (dalmia) [inline src/services/Functional/User.js:12]: every from "and save" should be written before the call to vuex before `return reAuthenticationCall`

File: src/App.vue [hunk 10]
```diff
@@ -115,6 +132,11 @@ export default {
       if (value) this.$Progress.start();
       else this.$Progress.finish();
     },
+    loginStatus(newValue, oldValue) {
+      // if user was logged in before but has been logged out now
+      // show a popup telling the user that they're logged out
+      if (oldValue && !newValue) this.showUserLoggedOutDialog();
```
Comment (dalmia) [inline src/App.vue:138]: would have to be done using isAuthentication.
Also, this is under watch - so, the old Value and newValue will necessarily be different. 
Given it is a boolean - as long as newValue is False, that check is enough.

File: src/App.vue [hunk 11]
```diff
@@ -148,10 +170,45 @@ export default {
...
+        text: "",
+        class: "",
+      };
+
+      // carry out the loginAgain action when dialog is closed
+      this.dialogAction = "loginAgain";
+      // show the dialogue
+      this.showDialogBox = true;
+      // route to the 404 page
+      this.$router.replace({ name: "404" });
```
Comment (dalmia) [inline src/App.vue:194]: why route them to 404?

File: src/App.vue [hunk 12]
```diff
@@ -148,10 +170,45 @@ export default {
...
+    },
+    dialogConfirmed() {
+      // invoked when the confirm button of the dialog box is clicked
+      this.showDialogBox = false;
+      this.dialogDescription = "";
```
Comment (dalmia) [inline src/App.vue:199]: why is this needed?

File: src/App.vue [hunk 13]
```diff
@@ -148,10 +170,45 @@ export default {
...
+      // close the dialog box when user clicks the button
+      if (this.dialogAction == "loginAgain") {
+        this.logoutUser();
+        this.showDialogBox = false;
+      }
+      // reset the dialog action value
+      this.dialogAction = "";
```
Comment (dalmia) [inline src/App.vue:206]: why is this needed?

File: src/App.vue [hunk 14]
```diff
@@ -148,10 +170,45 @@ export default {
...
+      // route to the 404 page
+      this.$router.replace({ name: "404" });
+    },
+    dialogConfirmed() {
+      // invoked when the confirm button of the dialog box is clicked
+      this.showDialogBox = false;
+      this.dialogDescription = "";
+      // close the dialog box when user clicks the button
+      if (this.dialogAction == "loginAgain") {
+        this.logoutUser();
```
Comment (dalmia) [inline src/App.vue:202]: I don't think the current implementation of logout user redirects the user back to the page they are coming from - it only directs to Home.

File: src/services/API/RootClient.js [hunk 15]
```diff
@@ -42,9 +60,38 @@ client.interceptors.request.use(
...
+
+    // If refresh token is invalid (400 BAD REQUEST) then log out the user
+    if (error.config.url == refreshTokenEndpoint && status === 400) {
+      store.dispatch("auth/logoutUser");
+      return Promise.reject(error);
+    }
+
+    // Handle expired/deleted access token here
+    if (status === 401) {
+      // re authenticate the user, and until that is happenining, pause all other
```
Comment (dalmia) [inline src/services/API/RootClient.js:85]:
```suggestion
      // reauthenticate the user, and until that is happenining, pause all other
```

File: src/services/API/RootClient.js [hunk 16]
```diff
@@ -42,9 +60,38 @@ client.interceptors.request.use(
...
+      // requests. Once the user is authenticated, release all the paused requests
+      // with the new token attached to the header
+      return UserFunctionalService.reAuthenticate(store).then(() => {
+        error.config.headers["Authorization"] = `
+            Bearer ${store.state.auth.accessToken.access_token}`;
+        return client.request(error.config);
+      });
```
Comment (dalmia) [inline src/services/API/RootClient.js:92]: didn't understand this part - how is it pausing the other requests?

Comment (dalmia) [reply src/services/API/RootClient.js:92]: Has this been tested for users set up with OTP?

File: src/services/API/RootClient.js [hunk 17]
```diff
@@ -42,9 +60,38 @@ client.interceptors.request.use(
 // handle errors in responses for API requests
 client.interceptors.response.use(
   (config) => {
+    // cancel the axios response if the user is not logged in
+    // and do it only for the requests that are not related to authentication
+    if (
+      store.state.auth.loginStatus == false &&
```
Comment (dalmia) [inline src/services/API/RootClient.js:66]:
```suggestion
      ! store.state.auth.loginStatus &&
```

File: src/App.vue [hunk 18]
```diff
@@ -132,10 +115,13 @@ export default {
       if (value) this.$Progress.start();
       else this.$Progress.finish();
     },
-    loginStatus(newValue, oldValue) {
+    isAuthenticated(value) {
       // if user was logged in before but has been logged out now
       // show a popup telling the user that they're logged out
-      if (oldValue && !newValue) this.showUserLoggedOutDialog();
+      if (!value) {
+        this.toast.error("You have been logged out!");
```
Comment (dalmia) [inline src/App.vue:122]: need to be localized

File: src/App.vue [hunk 19]
```diff
@@ -146,7 +132,7 @@ export default {
     logoutUser() {
       // logs out the user
       this.unsetAccessToken().then(() => {
-        this.$router.replace({ name: "Login" });
+        this.$router.push({ name: "Login" });
```
Comment (dalmia) [inline src/App.vue:135]: why `push` instead of `replace`?

File: src/store/modules/auth.js [hunk 20]
```diff
@@ -51,14 +49,11 @@ const actions = {
   setReAuthenticationState({ commit }, isReAuthenticating) {
     commit("setReAuthenticationState", isReAuthenticating);
   },
-  setReAuthenticationCall({ commit }, reAuthenticationCall) {
-    commit("setReAuthenticationCall", reAuthenticationCall);
+  setreAuthenticationPromise({ commit }, reAuthenticationPromise) {
```
Comment (dalmia) [inline src/store/modules/auth.js:52]:
```suggestion
  setReAuthenticationPromise({ commit }, reAuthenticationPromise) {
```

File: src/store/modules/auth.js [hunk 21]
```diff
@@ -51,14 +49,11 @@ const actions = {
...
+    commit("setreAuthenticationPromise", reAuthenticationPromise);
   },
-  unsetReAuthenticationCall({ commit }) {
-    commit("unsetReAuthenticationCall");
-  },
-  logoutUser({ commit }) {
-    commit("logoutUser");
+  unsetreAuthenticationPromise({ commit }) {
```
Comment (dalmia) [inline src/store/modules/auth.js:55]:
```suggestion
  unsetReAuthenticationPromise({ commit }) {
```

File: src/store/modules/auth.js [hunk 22]
```diff
@@ -92,17 +87,11 @@ const mutations = {
...
   },
-  setReAuthenticationCall(state, reAuthenticationCall) {
-    state.reAuthenticationCall = reAuthenticationCall;
-  },
-  unsetReAuthenticationCall(state) {
-    state.reAuthenticationCall = null;
-  },
-  logoutUser(state) {
-    state.loginStatus = false;
+  setreAuthenticationPromise(state, reAuthenticationPromise) {
```
Comment (dalmia) [inline src/store/modules/auth.js:90]:
```suggestion
  setReAuthenticationPromise(state, reAuthenticationPromise) {
```

File: src/store/modules/auth.js [hunk 23]
```diff
@@ -92,17 +87,11 @@ const mutations = {
...
+    state.reAuthenticationPromise = reAuthenticationPromise;
   },
-  loginUser(state) {
-    state.loginStatus = true;
+  unsetreAuthenticationPromise(state) {
```
Comment (dalmia) [inline src/store/modules/auth.js:93]:
```suggestion
  unsetReAuthenticationPromise(state) {x
```

File: src/services/API/RootClient.js [hunk 24]
```diff
@@ -76,20 +76,22 @@ client.interceptors.response.use(
 
     // If refresh token is invalid (400 BAD REQUEST) then log out the user
     if (error.config.url == refreshTokenEndpoint && status === 400) {
-      store.dispatch("auth/logoutUser");
-      return Promise.reject(error);
+      store.dispatch("auth/unsetAccessToken");
+      throw new axios.Cancel("User has been logged out");
```
Comment (dalmia) [inline src/services/API/RootClient.js:80]: why change this from `Promise.reject`?

File: src/services/API/RootClient.js [hunk 25]
```diff
@@ -63,7 +63,7 @@ client.interceptors.response.use(
     // cancel the axios response if the user is not logged in
     // and do it only for the requests that are not related to authentication
     if (
-      store.state.auth.loginStatus == false &&
+      store.getters["auth/isAuthenticated"] == false &&
```
Comment (dalmia) [inline src/services/API/RootClient.js:66]:
```suggestion
      !store.getters["auth/isAuthenticated"] &&
```

File: src/services/Functional/User.js [hunk 26]
```diff
@@ -2,33 +2,33 @@ import UserAPIService from "@/services/API/User.js";
...
+    // return the stored promise from the re-authentication call if re-authentication is in process
     if (store.state.auth.isReAuthenticating)
-      return store.state.auth.reAuthenticationCall;
+      return store.state.auth.reAuthenticationPromise;
 
-    // re authentication is in process
+    // set re-authentication as being in process
     store.dispatch("auth/setReAuthenticationState", true);
 
     // start the refreshAccessToken call and save that promise in the variable
```
Comment (dalmia) [inline src/services/Functional/User.js:12]: the line after "and save" should be below as mentioned before as well - I think it got missed.

File: src/store/plugins/usersWebSocketPlugin.js [hunk 27]
```diff
@@ -2,34 +2,43 @@ import { usersEndpoint } from "@/services/API/Endpoints.js";
...
 
 export default function createUsersWebSocket() {
-  return function connect(store) {
-    var UsersWebsocket = new WebSocket(
-      process.env.VUE_APP_BACKEND_WEBSOCKET +
-        usersEndpoint +
-        store.state.auth.userId
-    );
+  return function connect(store, unsubscribeCallback) {
+    if (unsubscribeCallback != null) unsubscribeCallback();
```
Comment (dalmia) [inline src/store/plugins/usersWebSocketPlugin.js:6]: comments

File: src/store/plugins/usersWebSocketPlugin.js [hunk 28]
```diff
@@ -2,34 +2,43 @@ import { usersEndpoint } from "@/services/API/Endpoints.js";
...
+      };
+
+      UsersWebsocket.onerror = (error) => {
+        console.log("Websocket - an error occurred. Closing socket");
+        console.log(error);
+        UsersWebsocket.close();
+      };
+    }
+
+    const unsubscribe = store.subscribe((mutation) => {
```
Comment (dalmia) [inline src/store/plugins/usersWebSocketPlugin.js:39]: comments

File: src/store/plugins/usersWebSocketPlugin.js [hunk 29]
```diff
@@ -2,34 +2,43 @@ import { usersEndpoint } from "@/services/API/Endpoints.js";
...
-        store.state.auth.userId
-    );
+  return function connect(store, unsubscribeCallback) {
+    if (unsubscribeCallback != null) unsubscribeCallback();
 
-    UsersWebsocket.onopen = () => {
-      console.log("WebSocket is open now.");
-    };
+    if (store.getters["auth/isAuthenticated"]) {
+      var UsersWebsocket = new WebSocket(
```
Comment (dalmia) [inline src/store/plugins/usersWebSocketPlugin.js:9]: define this before the if and assign value here.

File: src/store/plugins/usersWebSocketPlugin.js [hunk 30]
```diff
@@ -2,34 +2,43 @@ import { usersEndpoint } from "@/services/API/Endpoints.js";
 const RECONNECT_TIME = 2000;
```
Comment (dalmia) [inline src/store/plugins/usersWebSocketPlugin.js:2]: need to fetch user details in App.vue so that if their status / orgs was updated when they were away - the change reflects when they come back.

**Comment by dalmia** [comment]
Still need to figure out how to deploy new packages on docker without manual intervention

File: src/assets/locales/hi.js [hunk 31]
```diff
@@ -162,6 +162,7 @@ export default {
     generic: "कुछ दिक्कत आ गयी कृपया दोबारा कोशिश करें",
     internet: "कृपया अपने इंटरनेट कनेक्शन की जाँच करें",
     copying: "कापी करते वक़्त दिक्कत हुई ",
+    auto_logout: "आप लॉग आउट हो चुके हैं!",
```
Comment (dalmia) [inline src/assets/locales/hi.js:165]: "aapko logout kar dia gaya hain" would be the appropriate translation.

Comment (deepansh96) [reply src/assets/locales/hi.js:165]: I just used google translate
I think this makes sense
Anyway we'll get all these texts verified, so can park until that happens

File: src/store/plugins/usersWebSocketPlugin.js [hunk 32]
```diff
@@ -3,41 +3,54 @@ const RECONNECT_TIME = 2000;
...
         UsersWebsocket.close();
       };
     }
 
     const unsubscribe = store.subscribe((mutation) => {
+      // subscribe to store mutations
+
+      // if the user is set, make a new websocket connection and unsubscribe to the mutations
       if (mutation.type === "auth/setUser") connect(store, unsubscribe);
+      // if the user is unset, close the websocket
```
Comment (dalmia) [inline src/store/plugins/usersWebSocketPlugin.js:53]:
```suggestion
      // if the user has been unset, close the websocket
```

File: src/store/plugins/usersWebSocketPlugin.js [hunk 33]
```diff
@@ -3,41 +3,54 @@ const RECONNECT_TIME = 2000;
...
         console.log("Websocket - an error occurred. Closing socket");
         console.log(error);
         UsersWebsocket.close();
       };
     }
 
     const unsubscribe = store.subscribe((mutation) => {
+      // subscribe to store mutations
+
+      // if the user is set, make a new websocket connection and unsubscribe to the mutations
```
Comment (dalmia) [inline src/store/plugins/usersWebSocketPlugin.js:51]:
```suggestion
      // if the user has been newly set, make a new websocket connection and unsubscribe to the mutations of the previous websocket connection
```

File: src/store/plugins/usersWebSocketPlugin.js [hunk 34]
```diff
@@ -3,41 +3,54 @@ const RECONNECT_TIME = 2000;
 
 export default function createUsersWebSocket() {
   return function connect(store, unsubscribeCallback) {
+    // unsubscribe a store subscription made in the previous call
```
Comment (dalmia) [inline src/store/plugins/usersWebSocketPlugin.js:6]: either we can remove this or we need to change this - I don't this anyone reading this would understand what this line means.

File: src/services/Functional/User.js [hunk 35]
```diff
@@ -9,26 +9,28 @@ export default {
     // set re-authentication as being in process
     store.dispatch("auth/setReAuthenticationState", true);
 
-    // start the refreshAccessToken call and save that promise in the variable
-    // "reAuthenticationPromise". Until that promise is resolved, store this promise
-    // into vuex and return from the function.
+    // start the refreshAccessToken call
     const reAuthenticationPromise = UserAPIService.refreshAccessToken()
+      //  Save that promise in a variable called "reAuthenticationPromise"
```
Comment (dalmia) [inline src/services/Functional/User.js:14]: sorry my bad - this line should still be above.

File: src/services/API/RootClient.js [hunk 36]
```diff
@@ -42,9 +60,40 @@ client.interceptors.request.use(
 // handle errors in responses for API requests
 client.interceptors.response.use(
   (config) => {
+    // cancel the axios response if the user is not logged in
+    // and do it only for the requests that are not related to authentication
+    if (
+      !store.getters["auth/isAuthenticated"] &&
+      !authEndpoints.includes(config.config.url)
+    )
+      throw new axios.Cancel("User is not logged in");
```
Comment (dalmia) [inline src/services/API/RootClient.js:69]: i don't think I got the explanation for this + string translation was missed out here.

File: src/services/API/RootClient.js [hunk 37]
```diff
@@ -65,8 +75,11 @@ client.interceptors.response.use(
       // with the new token attached to the header
       return UserFunctionalService.reAuthenticate(store)
         .then(() => {
-          error.config.headers["Authorization"] = `
-            Bearer ${store.state.auth.accessToken.access_token}`;
+          var newAccessToken = store.state.auth.accessToken.access_token;
+          error.config.headers["Authorization"] = `Bearer ${newAccessToken}`;
+
+          if (error.config.url == userFromTokenEndpoint + "/")
```
Comment (dalmia) [inline src/services/API/RootClient.js:81]: one comment

File: src/services/API/RootClient.js [hunk 38]
```diff
@@ -52,10 +55,17 @@ client.interceptors.response.use(
...
     const status = error.response ? error.response.status : null;
 
-    // If refresh token is invalid (400 BAD REQUEST) then log out the user
+    // If refresh token is invalid (400 BAD REQUEST)
     if (error.config.url == refreshTokenEndpoint && status === 400) {
+      // unset the access token and log out the user
       store.dispatch("auth/unsetAccessToken");
-      return Promise.reject(error);
+
+      // set the re authentication status as false
```
Comment (dalmia) [inline src/services/API/RootClient.js:63]: reauthentication
